test/unit/math/mix/meta/eigen_utils.hpp

```diff
@@ -0,0 +1,298 @@
+          bool eigen_sparse_matrix_v, bool eigen_sparse_map_v, typename Scalar,
+          template <class...> class Checker>
+void test_eigen_sparse_matrix() {
+  using Eigen::EigenBase;
+  using Eigen::SparseCompressedBase;
+  using Eigen::SparseMapBase;
+  using Eigen::SparseMatrix;
+  using Eigen::SparseMatrixBase;
+  using SparseMatrixXd = SparseMatrix<Scalar>;
+  SparseMatrixXd C(10, 10);
+  SparseMatrixXd D(10, 10);
+  // scalars
+  EXPECT_FALSE((Checker<bool>::value));
+  EXPECT_FALSE((Checker<double>::value));
+  EXPECT_FALSE((Checker<int>::value));
+
+  // Sparse
+  EXPECT_TRUE((eigen_base_v == Checker<EigenBase<decltype(C * D)>>::value));
+  EXPECT_TRUE((eigen_sparse_compressed_v
+               == Checker<SparseCompressedBase<decltype(C * D)>>::value));
+  EXPECT_TRUE((eigen_sparse_matrix_v
+               == Checker<SparseMatrixBase<decltype(C * D)>>::value));
+  EXPECT_TRUE((eigen_sparse_matrix_v == Checker<decltype(C * D)>::value));
+  EXPECT_TRUE((eigen_sparse_matrix_v == Checker<decltype(C * D)>::value));
+  EXPECT_TRUE(
+      (eigen_sparse_map_v == Checker<SparseMapBase<decltype(C * D)>>::value));
+}
+
+/**
+ * Test type checking for dense matrices
+ * @tparam eigen_base_v logical for whether `Checker` should pass for
+ *   `EigenBase` types.
+ * @tparam eigen_dense_v logical for whether `Checker` should pass for
+ *   `DenseBase` types.
+ * @tparam eigen_matrix_v logical for whether `Checker` should pass for
+ *   `EigenMatrixBase` types.
+ * @tparam eigen_matrix_expr_v logical for whether `Checker` should pass for
+ *   `EigenMatrixBase` expression types.
+ * @tparam eigen_map_v logical for whether `Checker` should pass for
+ *   `Map` expression types.
+ * @tparam Scalar The scalar type of the Eigen object to check.
+ * @tparam Checker A struct taking in an expression template type and returning
+ *  a true if check is satisfied and false otherwise.
+ * @tparam EigenDims template parameter pack of `int` defining matrix
+ *  compile time dimensions.
+ */
+template <
+    bool eigen_base_v, bool eigen_dense_v, bool eigen_matrix_v,
+    bool eigen_array_v, bool eigen_map_v, template <class...> class Checker,
+    template <class, int...> class EigenType, typename Scalar, int... EigenDims>
+void test_eigen_dense() {
+  using Eigen::ArrayBase;
+  using Eigen::DenseBase;
+  using Eigen::EigenBase;
+  using Eigen::Map;
+  using Eigen::MatrixBase;
+  using test_mat = EigenType<Scalar, EigenDims...>;
+  test_mat A;
+  test_mat B;
+  // scalars
+  EXPECT_FALSE((Checker<bool>::value));
+  EXPECT_FALSE((Checker<double>::value));
+  EXPECT_FALSE((Checker<int>::value));
+
+  //  Dense Matrix Hierarchy
+  EXPECT_TRUE((eigen_base_v == Checker<EigenBase<test_mat>>::value));
+  EXPECT_TRUE((eigen_base_v == Checker<EigenBase<decltype(A * B)>>::value));
+  EXPECT_TRUE((eigen_dense_v == Checker<DenseBase<decltype(A * B)>>::value));
+  EXPECT_TRUE((eigen_matrix_v == Checker<MatrixBase<decltype(A * B)>>::value));
```

Comment:
Again checks on base classes. Instead I would suggest adding a check on `decltype((A * B).array())` and on plain `A`.

Reply:
See my comment above. I agree with a check on `.array()`, though that's also covered by the `ArrayBase` test

Reply:
Where is `ArrayBase` test? I can't find it. If I follow your logic it should be in this function?

Reply:
I see I was looking at subset of commits. This is fine.